ocrd_models/ocrd_models/ocrd_mets.py

```diff
@@ -310,19 +310,36 @@ def add_file(self, fileGrp, mimetype=None, url=None, ID=None, pageId=None, force
             raise ValueError("Invalid syntax for mets:file/@ID %s (not an xs:ID)" % ID)
         if not REGEX_FILE_ID.fullmatch(fileGrp):
             raise ValueError("Invalid syntax for mets:fileGrp/@USE %s (not an xs:ID)" % fileGrp)
+        log = getLogger('ocrd_models.ocrd_mets.remove_file_group')
         el_fileGrp = self._tree.getroot().find(".//mets:fileGrp[@USE='%s']" % (fileGrp), NS)
         if el_fileGrp is None:
             el_fileGrp = self.add_file_group(fileGrp)
-        mets_file = next(self.find_files(ID=ID), None)
-        if mets_file and not ignore:
-            if not force:
-                raise Exception("File with ID='%s' already exists" % ID)
-            mets_file.url = url
-            mets_file.mimetype = mimetype
-            mets_file.ID = ID
-            mets_file.pageId = pageId
-            mets_file.local_filename = local_filename
-        else:
+        mets_file = None
+        if not ignore:
+            if pageId:
+                mets_file = next(self.find_files(fileGrp=fileGrp, mimetype=mimetype, pageId=pageId), None)
+                if mets_file:
+                    if not force:
+                        # XXX should this be an exception?
```

Comment:
But if it is an exception here, we cannot use the second find query as a fallback. Unless we change the logic to not extend any existing OcrdFile like you propose above.

Reply:
That fallback is illegitimate just the same, though: If the ID already exists, it may be the file we want to overwrite, or it may not be (i.e. a clash by chance).

Reply:
Yes, that was not the right solution, please have another look at the changed behavior.